winch/codegen/src/codegen/context.rs

```diff
@@ -330,32 +294,82 @@ impl<'a> CodeGenContext<'a> {
         Self::spill_impl(&mut self.stack, &mut self.regalloc, &mut self.frame, masm);
     }
 
+    /// Prepares the compiler to emit an uncoditional jump to the
+    /// given destination branch.  This process involves:
+    /// * Balancing the machine stack pointer by popping it to
+    ///   match the destination branch.
+    /// * Updating the reachability state.
+    /// * Marking the destination frame as a destination target.
+    pub fn unconditional_jump<M, F>(&mut self, dest: &mut ControlStackFrame, masm: &mut M, mut f: F)
+    where
+        M: MacroAssembler,
+        F: FnMut(&mut M, &mut Self, &mut ControlStackFrame),
+    {
+        let (_, target_sp) = dest.original_stack_len_and_sp_offset();
+        // Invariant: The SP, must be greater or equal to the target
+        // SP, given that we haven't popped any results by this point
+        // yet. But it may happen in the callback.
+        assert!(masm.sp_offset() >= target_sp);
+        f(masm, self, dest);
+
+        // The following snippet, pops the stack pointer to ensure
+        // that it is correctly placed according to the expectations
+        // of the destination branch.
+        //
+        // This is done in the context of unconditional jumps, as the
+        // machine stack might be left unbalanced at the jump site,
+        // due to register spills. In this context unbalanced refers
+        // to possible extra space created at the jump site, which
+        // might cause invaid memory accesses. Note that in some cases
+        // the stack pointer offset might be already less than or
+        // equal to the original stack pointer offset registered when
+        // entering the destination control stack frame, which
+        // effectively means that when reaching the jump site no extra
+        // space was allocated similar to what would happen in a fall
+        // through in which we assume that the program has allocated
+        // and deallocated the right amount of stack space.
```

Comment:
I know this is pre-existing text (I might have even reviewed it the first time?) but for some reason I'm having trouble wrapping my head around the `current_sp < target_sp` case here. Is the argument basically that the Wasm stack invariants (proper typing of blocks, etc) will fix it up? Or is this handled at the use-cases of this function? Either way it would be helpful to point out concretely what piece of the codegen handles the popping.

Reply:
I've added a bit more of docs on how this could happen. In general the first assert in this function (above) should always be true, which is what I was after with this refactor, which makes it easier to reason about IMHO. But with that comment I'm noting that when the code reaches the the stack pointer balancing part, the SP might be less if during the invocation of the callback the caller preemptively handles return values via `CodeGenContext::pop_abi_results` (e.g. what would happen during the emission of `br`). Hope this clarifies things.